### DISCUSSION TITLE
Add elements to initialized custom collection

### DISCUSSION THREAD

**Discussion by nlisker** [FasterXML/jackson-databind#4756] [Q&A]
(Copied from the mailing list)

My class contains an initialized custom collection

```java
class MyClass {

    MyList<Integer> myList = MyList.create();
    //getter
}
```

where `MyList` is a `j.u.List`. When deserializing, I get an exception saying that there is no known deserializer for `MyList` ("Cannot find a deserializer for non-concrete List type"), which is generally understandable, but I already initialized it, so there's no need for one. Instead, I just want to add the items to the list.
To demonstrate, I can add a private setter inside the class:

```java
private void setMyList(List<Integer> ints) { ints.forEach(MyList::add); }
```

this works, but I don't think it should be needed. Specifically, Jackson should not create any list. It needs to deserialize the values of the items anyway; then instead of creating a list, add them to the existing one. It's a `List`, so it should know to call 'add' ('put' for map).

I know there is `@JsonMerge`, and I thought that that's what I needed:

> merging meaning that the current value is first accessed (with a getter or field) and then modified with incoming data. If merging is not used assignment happens without considering current state.
...
For [Collection](https://docs.oracle.com/javase/6/docs/api/java/util/Collection.html?is-external=true) and Arrays, merging is done by appending incoming data into contents of existing Collection/array

but just adding it to the field doesn't work, I still get the same exception. Am I not understanding `@JsonMerge` or am I doing something wrong?

Thanks!

**Comment by nlisker** [comment]
From the link you gave.

**Reply by JooHyukKim** [reply]
This is one of passing test case from Jackson. And there is more you can look into.    
If same problem occurs, try at least 2.18 version of Jackson.

```
    static class MergedList {
        @JsonMerge
        List<String> values = new ArrayList<>();

        {
            values.add("a");
        }
    }

    @Test
    public void testListMerging() throws Exception {
        MergedList w = MAPPER.readValue(a2q("{'values':['x']}"), MergedList.class);
        assertEquals(2, w.values.size());
        assertTrue(w.values.contains("a"));
        assertTrue(w.values.contains("x"));
    }
```

**Reply by nlisker** [reply]
What is `a2q` here and how was `MAPPER` constructed? Where is this test from?

Besides, this example does not conform to the problem I laid out because it doesn't involve a custom collection. Here is a valid example:

```java
import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonMerge;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.json.JsonMapper;

interface MyList<T> extends List<T> {}

class MergeList {

	// dummy implementation just for the example to compile
	private static class MyArrayList<T> extends ArrayList<T> implements MyList<T> {}
	static <T> MyList<T> create() {
		return new MyArrayList<T>();
	}

	@JsonMerge
	@JsonProperty
	MyList<String> values = create();

	{
		values.add("a");
	}

	public static void main(String[] args) throws Exception {
		JsonMapper MAPPER = JsonMapper.builder().build();
		MergeList mergeList = new MergeList();
		var string = MAPPER.writeValueAsString(mergeList);
		System.out.println(string);
		MergeList w = MAPPER.readValue(("{\"values\":[\"x\"]}"), MergeList.class);
		System.out.println(w.values);
	}
}
```

Throws
```
Exception in thread "main" com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot find a deserializer for non-concrete Collection type [collection type; class example.MyList, contains [simple type, class java.lang.String]]
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 1]
	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:67)
	at com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1888)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:321)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:284)
	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:174)
	at com.fasterxml.jackson.databind.DeserializationContext.findNonContextualValueDeserializer(DeserializationContext.java:659)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.resolve(BeanDeserializerBase.java:552)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:347)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:284)
	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:174)
	at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer(DeserializationContext.java:669)
	at com.fasterxml.jackson.databind.ObjectMapper._findRootDeserializer(ObjectMapper.java:5048)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4918)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3860)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3828)
	at example.MergeList.main(MergeList.java:33)
```
But I don't need Jackson to find a deserializer for `MyList`, I already created it.

**Reply by JooHyukKim** [reply]
How did you declare `MyList`?
Seems like test passes when `values` field declared like...

```java
@JsonMerge
@JsonProperty
MyList<String> values = create();
```

... this (try it).

**Reply by JooHyukKim** [reply]
It seems like `BasicDeserializerFactory._mapAbstractCollectionType()` is returning `null` because `ContainerDefaultMappings.findCollectionFallback()` looks for direct interfaces and not extensions (here we have `MyList`).

```java
    protected CollectionType _mapAbstractCollectionType(JavaType type, DeserializationConfig config)
    {
        final Class<?> collectionClass = ContainerDefaultMappings.findCollectionFallback(type);
        if (collectionClass != null) {
            return (CollectionType) config.getTypeFactory()
                    .constructSpecializedType(type, collectionClass, true);
        }
        return null;
    }
```

Is it possible to make `ContainerDefaultMappings.findCollectionFallback()` also look for input type's interfaces and not just type's direct class, @cowtowncoder ?

**Reply by nlisker** [reply]
> How did you declare `MyList`?

It's in the code I posted.

> Seems like test passes when values field declared like...
> 
> ```
> @JsonMerge
> @JsonProperty
> MyList<String> values = create();
> ```
> ... this (try it).

This is a copy-paste of my example. It does not pass.

**Reply by JooHyukKim** [reply]
No I mean, is it this?

```java
public interface MyList extends List { }
```

**Reply by nlisker** [reply]
> It seems like `BasicDeserializerFactory._mapAbstractCollectionType()` is returning `null` because `ContainerDefaultMappings.findCollectionFallback()` looks for direct interfaces and not extensions (here we have `MyList`).
> 
> ```java
>     protected CollectionType _mapAbstractCollectionType(JavaType type, DeserializationConfig config)
>     {
>         final Class<?> collectionClass = ContainerDefaultMappings.findCollectionFallback(type);
>         if (collectionClass != null) {
>             return (CollectionType) config.getTypeFactory()
>                     .constructSpecializedType(type, collectionClass, true);
>         }
>         return null;
>     }
> ```
> 
> Is it possible to make `ContainerDefaultMappings.findCollectionFallback()` also look for input type's interfaces and not just type's direct class, @cowtowncoder ?

More than that, why is it looking for a deserializer at all? It only needs to deserialize the values of the collection (`String` etc.) and add them to an already existing collection. There's no need to create a new collection, and in fact, it can be detrimental to do so.

I will reiterate that there *is* a solution with a custom setter as I've shown above. My thoughts were that `@JsonMerge` would do it for me, and according to its docs it should.

**Reply by nlisker** [reply]
> No I mean, is it this?
> 
> ```java
> public interface MyList extends List { }
> ```

Yes, this is it. Copy-paste the whole code snippet and try it. It's the simplest minimal example I could create based on the test you gave.

**Reply by JooHyukKim** [reply]
> More than that, why is it looking for a deserializer at all? It only needs to deserialize the values of the collection (String etc.).

That is what `StringCollectionDeserializer` does exactly. It simply adds to existing list. Refer to its `public Collection<String> deserialize(JsonParser, DeserializationContext, Collection<String>)` method.


> I will reiterate that there is a solution with a custom setter as I've shown above. My thoughts were that @JsonMerge would do it for me, and according to its docs it should.

I think your solution is more of a work-around. It should probably work without a setter.

**Comment by nlisker** [comment]
Submitted https://github.com/FasterXML/jackson-databind/issues/4783.

**Reply by JooHyukKim** [reply]
Thanks!
